File: src/utils/splash.ts
/**
 * 开屏动画（splashscreen）就绪通知
 *
 * 启动时序：
 * 1. Tauri 先创建开屏窗口（splash.html 动画），main 窗口 visible:false 后台加载
 * 2. Vue app 挂载后调用 notifyFrontendReady()
 * 3. 为保证动画至少展示一段时间（避免一闪而过），不足最短时长则延迟到最短时长再通知
 * 4. 后端 frontend_ready 命令关闭开屏窗口、显示主窗口
 *
 * 开屏页与动画兜底见 public/splash.html（4.6s 未收到前端信号时自兜底调用同一命令）。
 */

import { invoke } from '@tauri-apps/api/core'

/** 开屏动画完整展示时长（ms）：与 splash.js 动画总时长（约 4.6s）对齐，避免动画播一半被切走 */
const SPLASH_MIN_MS = 4600

/** 模块加载时刻（约等于 WebView 加载 JS bundle 的时刻），作为最短展示计时起点 */
const startedAt = Date.now()

/**
 * 通知后端开屏动画就绪：至少展示 SPLASH_MIN_MS 后关闭开屏窗口、显示主窗口
 *
 * 浏览器预览环境（非 Tauri）invoke 会失败，静默忽略。
 */
export function notifyFrontendReady(): void {
  const elapsed = Date.now() - startedAt
  const wait = Math.max(0, SPLASH_MIN_MS - elapsed)
  window.setTimeout(() => {
    invoke('frontend_ready').catch(() => {
      // 浏览器预览环境无该命令，忽略即可
    })
  }, wait)
}

File: src-tauri/src/splash.rs
//! 开屏动画（splashscreen）窗口控制
//!
//! 启动时序：
//! 1. Tauri 先创建 `splashscreen` 窗口（加载 `splash.html` 开屏动画，无边框透明居中）
//! 2. `main` 窗口以 `visible: false` 后台加载（Rust 初始化 + WebView 加载大 bundle 期间
//!    用户看到的是开屏动画而非卡顿白屏）
//! 3. 前端 Vue app 就绪后（或开屏动画播完兜底）调用 `frontend_ready`
//! 4. 本命令关闭开屏窗口、显示并聚焦主窗口
//!
//! 开屏页位于 `public/splash.html`（Vite 构建时复制到 dist，dev 模式经 devUrl 提供），

use crate::log_info;
use tauri::Manager;

/// 前端就绪：关闭开屏窗口并显示主窗口
///
/// 幂等：开屏窗口已关闭 / 主窗口已显示时静默跳过，重复调用无副作用。
/// 调用方：前端 Vue 就绪后主动调用；开屏页动画播完后也做兜底调用。
#[tauri::command]
pub fn frontend_ready(app: tauri::AppHandle) -> Result<(), String> {
    if let Some(splash) = app.get_webview_window("splashscreen") {
        let _ = splash.close();
        log_info!("[Splash] 开屏窗口已关闭");
    }
    if let Some(main_window) = app.get_webview_window("main") {
        let _ = main_window.show();
        let _ = main_window.unminimize();
        let _ = main_window.set_focus();
        log_info!("[Splash] 主窗口已显示");
    }
    Ok(())
}
